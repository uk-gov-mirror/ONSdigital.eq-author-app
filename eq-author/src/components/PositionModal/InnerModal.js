import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { colors, radius } from "constants/theme";

import Truncated from "components/Truncated";
import ItemSelectModal from "components/ItemSelectModal";
import ItemSelect from "components/ItemSelectModal/ItemSelect";
import Icon from "assets/icon-select.svg";

const Label = styled.label`
  display: block;
  font-size: 1em;
  font-weight: bold;
  margin-bottom: 0.25rem;
  margin-top: 1.25rem;
`;

const Trigger = styled.button.attrs({ type: "button" })`
  width: 100%;
  font-size: 1em;
  padding: 0.5rem;
  padding-right: 2em;
  background: ${colors.white} url("${Icon}") no-repeat right center;
  border: solid 1px ${colors.borders};
  text-align: left;
  border-radius: ${radius};
  color: ${colors.black};

  &:focus {
    box-shadow: 0 0 0 3px ${colors.tertiary}, inset 0 0 0 1px ${colors.primary};
    outline: none;
  }
`;

export const InnerModal = ({
  id,
  title,
  isOpen,
  onClick,
  onClose,
  onConfirm,
  onChange,
  children,
  selected,
  displayName,
}) => {
  return (
    <>
      <Label htmlFor={id}>{title}</Label>
      <Trigger id={id} onClick={onClick}>
        <Truncated>{displayName || "Select"}</Truncated>
      </Trigger>
      <ItemSelectModal
        title={title}
        data-test={`${title.toLowerCase()}-select-modal`}
        isOpen={isOpen}
        onClose={onClose}
        onConfirm={onConfirm}
      >
        <ItemSelect
          data-test="section-item-select"
          name={title.toLowerCase()}
          value={selected?.id || selected}
          onChange={({ value }) => onChange(value)}
        >
          {children}
        </ItemSelect>
      </ItemSelectModal>
    </>
  );
};

InnerModal.propTypes = {
  id: PropTypes.string,
  title: PropTypes.string,
  isOpen: PropTypes.bool,
  onClick: PropTypes.func,
  onClose: PropTypes.func,
  onConfirm: PropTypes.func,
  onChange: PropTypes.func,
  children: PropTypes.node,
  selected: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      id: PropTypes.string,
      displayName: PropTypes.string,
      position: PropTypes.number,
    }),
  ]).isRequired,
  displayName: PropTypes.string,
};
