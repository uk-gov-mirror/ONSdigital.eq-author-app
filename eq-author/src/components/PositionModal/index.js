import React, { useState, useEffect, useRef } from "react";
import PropTypes from "prop-types";
import { uniqueId } from "lodash";
import styled from "styled-components";
import { InnerModal } from "./InnerModal";
import { Option } from "components/ItemSelectModal/ItemSelect";

const Indent = styled(Option)`
  margin-left: ${({ indent }) => (indent ? 1 : 0)}em;
`;

const PositionModal = ({ title, options, onMove, selected, onChange }) => {
  const positionButtonId = uniqueId("PositionModal");
  const [isOpen, setIsOpen] = useState(false);

  const previousIndex = options.findIndex(({ id }) => id === selected?.id);
  const previousPosition = useRef(previousIndex > -1 ? previousIndex : 0);

  const [{ position, item }, setOption] = useState({
    position: previousPosition.current,
    item: options[previousPosition.current],
  });

  useEffect(() => {
    // resets the position of the selected item when changing sections
    const previousIndex = options.findIndex(({ id }) => id === selected?.id);
    previousPosition.current = previousIndex > -1 ? previousIndex : 0;
    setOption((prev) => ({ ...prev, position: previousPosition.current }));
  }, [options, selected]);

  const orderedOptions = options.filter(({ id }) => id !== selected?.id);
  selected.parentEnabled = item?.parentEnabled;
  orderedOptions.splice(position, 0, selected);

  const handleClose = () => {
    setIsOpen(false);
    setOption({
      position: previousPosition.current,
      item: orderedOptions[previousPosition.current],
    });
  };

  const handleChange = (value) => {
    const count =
      orderedOptions[value]?.__typename === "Folder" && value - position >= 0
        ? orderedOptions.filter(
            ({ parentId }) => parentId === orderedOptions[value].id
          ).length
        : 0;
    setOption({
      position: parseInt(value, 10) + count,
      item: orderedOptions[value],
    });
  };

  const handleConfirm = (e) => {
    e.preventDefault();
    // item is the element in the modal you just clicked
    const { parentId = null } = item;

    let positionCalculation;

    if (parentId) {
      // get pages in target folder + selected item
      positionCalculation = orderedOptions
        .filter(
          ({ parentId: itemId, id }) =>
            parentId === itemId || id === selected.id
        )
        .findIndex(({ id }) => id === selected.id);
    } else {
      // remove all nested pages
      positionCalculation = orderedOptions
        .filter(({ parentId }) => !parentId)
        .findIndex(({ id }) => id === selected.id);
    }

    if (onMove) {
      onMove({
        ...item,
        folderId: parentId,
        position: positionCalculation,
      });
    }

    setIsOpen(false);
  };

  return (
    <div data-test="position-modal">
      <InnerModal
        id={positionButtonId}
        title={title}
        isOpen={isOpen}
        onClick={() => setIsOpen(true)}
        onClose={handleClose}
        onConfirm={handleConfirm}
        onChange={onChange || handleChange}
        selected={String(position)}
        displayName={selected.displayName}
      >
        {orderedOptions.map(({ displayName, parentEnabled }, i) => (
          <Indent
            data-test={`option-${i}`}
            key={i}
            value={String(i)}
            indent={parentEnabled ? parentEnabled.toString() : undefined}
          >
            {displayName}
          </Indent>
        ))}
      </InnerModal>
    </div>
  );
};

PositionModal.propTypes = {
  title: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      displayName: PropTypes.string.isRequired,
      folderId: PropTypes.string,
      parentEnabled: PropTypes.bool,
    })
  ).isRequired,
  onMove: PropTypes.func.isRequired,
  selected: PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string,
    position: PropTypes.number,
  }).isRequired,
  onChange: PropTypes.string,
};

export default PositionModal;
