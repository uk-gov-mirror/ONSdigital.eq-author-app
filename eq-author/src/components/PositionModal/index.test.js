import React from "react";
import PositionModal from "./";
import {
  render,
  fireEvent,
  screen,
  waitForElementToBeRemoved,
} from "tests/utils/rtl";
import { buildSections } from "tests/utils/createMockQuestionnaire";

describe("PositionModal", () => {
  const options = buildSections({ sectionCount: 5 });
  const onMove = jest.fn();
  const setup = (props = {}) =>
    render(
      <PositionModal
        title="Section"
        options={options}
        isOpen
        onClose={jest.fn()}
        onMove={onMove}
        selected={options[0]}
        {...props}
      />
    );

  beforeEach(() => setup());

  it("should render", () => {
    expect(screen.getByTestId("section-position-modal")).toBeVisible();
  });

  it("should open when clicked", () => {
    fireEvent.click(screen.getByText(/Section 1/));
    expect(screen.getByTestId("section-item-select")).toBeVisible();
  });

  it("should close on select", async () => {
    fireEvent.click(screen.getByText(/Section 1/));
    fireEvent.click(screen.getByText(/Select/));
    await waitForElementToBeRemoved(() => screen.getByText(/Select/));
    expect(screen.queryByText(/Select/)).not.toBeInTheDocument();
  });

  it("should close on cancel", async () => {
    fireEvent.click(screen.getByText(/Section 1/));
    fireEvent.click(screen.getByText(/Cancel/));
    await waitForElementToBeRemoved(() => screen.getByText(/Cancel/));
    expect(screen.queryByText(/Cancel/)).not.toBeInTheDocument();
  });

  it("should update selected on change", () => {
    fireEvent.click(screen.getByText(/Section 1/));
    fireEvent.click(screen.getByText(/Section 2/));
    fireEvent.click(screen.getByText(/Select/));
    // await waitForElementToBeRemoved(() => screen.getByText(/Select/));
    expect(onMove).toHaveBeenCalledWith(
      expect.objectContaining({
        displayName: "Section 2",
      })
    );
  });

  it("should reset the position if modal is closed", async () => {
    fireEvent.click(screen.getByText(/Section 1/));
    fireEvent.click(screen.getByText(/Section 2/));
    fireEvent.click(screen.getByText(/Cancel/));
    await waitForElementToBeRemoved(() => screen.getByText(/Cancel/));

    expect(screen.getByText(/Section 1/)).toBeVisible();
    expect(screen.queryByText(/Section 2/)).not.toBeInTheDocument();
  });

  describe("Positioning behaviours", () => {
    let options;
    function openModalState(selected = 0) {
      const utils = setup({
        title: "Page",
        options,
        selected: options[selected],
      });
      fireEvent.click(screen.getByText(/question 1/));
      return { ...utils };
    }
    beforeEach(() => {
      options = [
        {
          id: "question-1",
          displayName: "question 1",
          parentEnabled: false,
        },
        {
          id: "Folder-1",
          displayName: "Folder 1",
          __typename: "Folder",
        },
        {
          id: "question-2",
          displayName: "question 2",
          parentEnabled: true,
          parentId: "Folder-1",
        },
        {
          id: "question-3",
          displayName: "question 3",
          parentEnabled: true,
          parentId: "Folder-1",
        },
        {
          id: "question-4",
          displayName: "question 4",
          parentEnabled: false,
        },
        {
          id: "Folder-2",
          displayName: "Folder 2",
          __typename: "Folder",
        },
        {
          id: "question-5",
          displayName: "question 5",
          parentEnabled: true,
          parentId: "Folder-2",
        },
        {
          id: "question-6",
          displayName: "question 6",
          parentEnabled: true,
          parentId: "Folder-2",
        },
      ];
      openModalState();
    });

    it("should jump length of folder when clicking on a folder", () => {
      const first = screen.getAllByText(/question 1/)[1].textContent;
      const firstOption = screen.getAllByTestId("options")[0].textContent;
      expect(first).toEqual(firstOption);

      fireEvent.click(screen.getByText(/Folder 1/));

      expect(screen.getAllByTestId("options")[3].textContent).toEqual(first);
    });

    // should test parentId here

    // it("should go into the correct position in folder going down", () => {
    //   const wrapper = openModalState();
    //   const value = 2;
    //   getItemSelect(wrapper).simulate("change", { value });
    //   expect(getItemSelect(wrapper).prop("value")).toBe(value.toString());
    //   expect(
    //     wrapper.find("PositionModal__Indent").getElements()[value].props
    //       .children
    //   ).toEqual(options[0].displayName);
    //   expect(
    //     wrapper.find("PositionModal__Indent").getElements()[value].props.indent
    //   ).toBeTruthy();
    // });

    // it("should go into the correct position in folder going up", () => {
    //   const selected = 4;
    //   const wrapper = openModalState(selected);
    //   const value = 2;
    //   getItemSelect(wrapper).simulate("change", { value });
    //   expect(getItemSelect(wrapper).prop("value")).toBe(value.toString());

    //   expect(
    //     wrapper.find("PositionModal__Indent").getElements()[value].props
    //       .children
    //   ).toEqual(options[selected].displayName);
    // });

    // it("should go to the correct position when selecting the same value twice", () => {
    //   const wrapper = openModalState();
    //   const value = 1;
    //   getItemSelect(wrapper).simulate("change", { value });

    //   getItemSelect(wrapper).simulate("change", { value: 0 });

    //   expect(getItemSelect(wrapper).prop("value")).toBe((0).toString());

    //   expect(
    //     wrapper.find("PositionModal__Indent").getElements()[0].props.children
    //   ).toEqual(options[0].displayName);
    // });

    // it("should move between folders correctly", () => {
    //   const initial = 2;
    //   const wrapper = openModalState(initial);
    //   const after = 6;
    //   getItemSelect(wrapper).simulate("change", { value: after });

    //   expect(getItemSelect(wrapper).prop("value")).toBe(after.toString());

    //   expect(
    //     wrapper.find("PositionModal__Indent").getElements()[after].props
    //       .children
    //   ).toEqual(options[initial].displayName);
    // });

    it("should display the correct styling", () =>
      expect(screen.getAllByTestId("options")[3]).toHaveStyleRule(
        "margin-left",
        "1em"
      ));
  });
});
