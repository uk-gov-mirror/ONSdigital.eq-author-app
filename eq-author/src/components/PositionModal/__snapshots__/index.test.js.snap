// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionModal should render 1`] = `
<div
  data-test="position-modal"
>
  <InnerModal
    displayName="Section 1"
    id="PositionModal1"
    isOpen={false}
    onChange={[Function]}
    onClick={[Function]}
    onClose={[Function]}
    onConfirm={[Function]}
    selected="0"
  >
    <PositionModal__Indent
      data-test="option-0"
      key="0"
      value="0"
    >
      Section 1
    </PositionModal__Indent>
    <PositionModal__Indent
      data-test="option-1"
      key="1"
      value="1"
    >
      Section 2
    </PositionModal__Indent>
    <PositionModal__Indent
      data-test="option-2"
      key="2"
      value="2"
    >
      Section 3
    </PositionModal__Indent>
    <PositionModal__Indent
      data-test="option-3"
      key="3"
      value="3"
    >
      Section 4
    </PositionModal__Indent>
    <PositionModal__Indent
      data-test="option-4"
      key="4"
      value="4"
    >
      Section 5
    </PositionModal__Indent>
  </InnerModal>
</div>
`;
